File: src/builtin/super.ts
import Environment from "../runtime/environment";
import { Expression } from "../expression";
import { InternalType } from "../type";
import { evaluate } from "../util";
import { createSuper } from "../factory";
import FuncBody from "./funcBody";

/**
 * built-in `super()` function
 * @param args accept two arguments (required)
 * @param env
 * @returns a super object
 */
const super_: FuncBody = (
  args: Expression[],
  env: Environment
): InternalType => {
  const classObj = evaluate(args[0], env);
  const self = evaluate(args[1], env);
  return createSuper({ classObj, instance: self });
};

export default super_;

File: script/testfile/super.py
class A:
    def __init__(self, n):
        self.m = n * 2
        pass


class B(A):
    def __init__(self, n):
        s = super(B, self)
        s.__init__(n)
        self.n = n


class C(B):
    def __init__(self, n):
        s = super(C, self)
        s.__init__(n)


c = C(7)
print(c.n == 7 and c.m == 14)
